### ISSUE TITLE
[amplify_secure_storage] PlatformException(InvalidProtocolBufferException) in Amplify.configure

### ISSUE THREAD

**Issue by asaarnak** [aws-amplify/amplify-flutter#2500]
Exception during Amplify plugin configure.
Dependencies with latest transitive dependencies.
amplify_flutter: ^1.0.0-next.2
amplify_auth_cognito: ^1.0.0-next.2
  
https://github.com/aws-amplify/amplify-flutter/blob/9e7e9eee3b28ba02473a8d7ce59d1d03a5d2b439/packages/secure_storage/amplify_secure_storage/android/src/main/kotlin/com/amazonaws/amplify/amplify_secure_storage/amplify_secure_storage/EncryptedKeyValueRepository.kt#L38

Stacktraces that are related.

```
AuthPreconditionException(message: Credential store has error. Re-load before continuing., recoverySuggestion: null, underlyingException: null)
```

```
AuthException(message: PlatformException(c0, com.google.crypto.tink.shaded.protobuf.c0: Protocol message contained an invalid tag (zero)., Cause: null, Stacktrace: com.google.crypto.tink.shaded.protobuf.c0: Protocol message contained an invalid tag (zero).
	at com.google.crypto.tink.shaded.protobuf.z.J(Unknown Source:75)
	at com.google.crypto.tink.shaded.protobuf.z.G(Unknown Source:2)
	at xa.c0.W(Unknown Source:2)
	at ua.d.read(Unknown Source:8)
	at qa.b.a(Unknown Source:0)
	at ua.a$b.e(Unknown Source:27)
	at ua.a$b.f(Unknown Source:0)
	at ua.a$b.d(Unknown Source:11)
	at androidx.security.crypto.a.b(Unknown Source:50)
	at com.amazonaws.amplify.amplify_secure_storage.amplify_secure_storage.EncryptedKeyValueRepository.getSharedPreferences(Unknown Source:14)
	at com.amazonaws.amplify.amplify_secure_storage.amplify_secure_storage.EncryptedKeyValueRepository.get(Unknown Source:5)
	at com.amazonaws.amplify.amplify_secure_storage.amplify_secure_storage.AmplifySecureStoragePlugin.read(Unknown Source:19)
	at com.amazonaws.amplify.amplify_secure_storage.amplify_secure_storage.Messages$AmplifySecureStorageApi.lambda$setup$0(Unknown Source:30)
	at com.amazonaws.amplify.amplify_secure_storage.amplify_secure_storage.Messages$AmplifySecureStorageApi.c(Unknown Source:0)
	at com.amazonaws.amplify.amplify_secure_storage.amplify_secure_storage.a.a(Unknown Source:2)
	at io.flutter.plugin.common.a$b.a(Unknown Source:17)
	at se.c.l(Unknown Source:18)
	at se.c.m(Unknown Source:20)
	at se.c.i(Unknown Source:0)
	at se.b.run(Unknown Source:12)
	at se.c$h.d(Unknown Source:20)
	at se.c$h.e(Unknown Source:0)
	at se.c$h.b(Unknown Source:0)
	at se.d.run(Unknown Source:2)
	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1137)
	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:637)
	at java.lang.Thread.run(Thread.java:1012)
, null), recoverySuggestion: null, underlyingException: null). Error thrown null.
       at AmplifyAuthCognitoDart.configure(auth_plugin_impl.dart:302)
       at AmplifyAuthCognito.configure(auth_plugin_impl.dart:138)
       at Future.wait.<fn>(wait.java)
       at AmplifyHybridImpl.configurePlatform(hybrid_impl.dart:35)
       at AmplifyClass.configure(amplify_class.dart:107)
```

```
       at StateMachine._emitError(state_machine.dart:197)
       at StateMachine._listenForEvents(state_machine.dart:173)
       at AmplifySecureStorageApi.read(messages.android.g.dart:53)
       at CredentialStoreStateMachine.getVersion(credential_store_state_machine.dart:67)
       at CredentialStoreStateMachine.onMigrateLegacyCredentialStore(credential_store_state_machine.dart:281)
       at CredentialStoreStateMachineBase.resolve(credential_store_state_machine_base.dart:45)
```

```
       at StateMachine._emitError(state_machine.dart:197)
       at StateMachine._listenForEvents(state_machine.dart:173)
       at CredentialStoreStateMachine.getCredentialsResult(credential_store_state_machine.dart:59)
       at FetchAuthSessionStateMachine.onFetchAuthSession(fetch_auth_session_state_machine.dart:179)
       at FetchAuthSessionStateMachineBase.resolve(fetch_auth_session_state_machine_base.dart:37)
       at StateMachine._listenForEvents(state_machine.dart:171)
```

```
       at StateMachine._emitError(state_machine.dart:197)
       at StateMachine._listenForEvents(state_machine.dart:173)
       at FetchAuthSessionStateMachine.getLatestResult(fetch_auth_session_state_machine.dart:68)
       at SignInStateMachine._saveAuthResult(sign_in_state_machine.dart:636)
       at SignInStateMachine._processChallenge(sign_in_state_machine.dart:748)
       at SignInStateMachine.resolve(sign_in_state_machine.dart:858)
       at StateMachine._listenForEvents(state_machine.dart:171)
```

**Comment by asaarnak** [comment]
Phone: SM N986B
After clearing app cache the exception still occurred.
After clearing app data the exception is gone.

**Comment by asaarnak** [comment]
Maybe related to this dependency version:
https://github.com/google/tink/issues/413
https://issuetracker.google.com/issues/210008257
https://github.com/aws-amplify/amplify-flutter/blob/9e7e9eee3b28ba02473a8d7ce59d1d03a5d2b439/packages/secure_storage/amplify_secure_storage/android/build.gradle#L50

**Comment by Jordan-Nelson** [comment]
Hello @asaarnak - Thanks for opening the issue and providing the links to the other issues. I agree that this looks related to the other issues you linked.

Are you able to reproduce this consistently?

**Comment by asaarnak** [comment]
Nope, it has happened only once and only for 1 user.

**Comment by Jordan-Nelson** [comment]
Okay thanks. I think https://issuetracker.google.com/issues/164901843 is also reporting the same issue (looking at the stack trace provided for v1.1.0-alpha02 in the issue).

**Comment by asaarnak** [comment]
This happened again for one of our tester that downloaded and installed app from APK, we have no errors from google play installs on production.

**Comment by asaarnak** [comment]
On Android 3rd time this happened, we don't know who this was or how he install apk or google play.

Also noticed similar issue on iOS crash logs.
Should i create new task for iOS ?
```stacktrace
0  ???                            0x0 AmplifySecureStorageCupertino._add + 146 (amplify_secure_storage_cupertino.dart:146)
1  ???                            0x0 AmplifySecureStorageCupertino.write.<fn> + 65 (amplify_secure_storage_cupertino.dart:65)
2  ???                            0x0 (null).using + 124 (arena.dart:124)
3  ???                            0x0 AmplifySecureStorageCupertino.write + 63 (amplify_secure_storage_cupertino.dart:63)
4  ???                            0x0 AmplifySecureStorageDartMixin.write + 42 (amplify_secure_storage_mixin.vm.dart:42)
5  ???                            0x0 SecureStorageWorker.run + 73 (secure_storage_worker.dart:73)
6  ???                            0x0 (null)._run + 14 (secure_storage_worker.worker.vm.dart:14)
```

```stacktrace
SecureStorageException(
  message: An unknown exception occurred.,
  recoverySuggestion: An unknown exception occurred. Please take a look at https://github.com/aws-amplify/amplify-flutter/issues to see if there are any existing issues that match your scenario, and file an issue with the details of the bug if there isn't.,
  underlyingException: SecurityFrameworkError: {code: -25308, message: No error string is available.}),
  stackTrace=
  #0 AmplifySecureStorageCupertino._add (package:amplify_secure_storage_dart/src/platforms/amplify_secure_storage_cupertino.dart:146) 
  #1 AmplifySecureStorageCupertino.write.<anonymous closure> (package:amplify_secure_storage_dart/src/platforms/amplify_secure_storage_cupertino.dart:65) 
  #2 using (package:ffi/src/arena.dart:124) 
  #3 AmplifySecureStorageCupertino.write (package:amplify_secure_storage_dart/src/platforms/amplify_secure_storage_cupertino.dart:63) 
  #4 AmplifySecureStorageDartMixin.write (p
```

**Comment by Jordan-Nelson** [comment]
Hello @asaarnak - If you could open a new issue for the iOS crashes you are seeing that would be great.

I believe we have a fix for the Android issue. I just need to reproduce it first to be sure that it will resolve it.

**Comment by asaarnak** [comment]
Thanks! Created #2535

**Comment by Jordan-Nelson** [comment]
I was able to reproduce this on a physical device with the following steps:
- Run the app and login
- Close the app
- Force a device backup from Settings
- Uninstall the app
- Run the app and login

I am working on a fix.

**Comment by alexandrudrosca** [comment]
I also started to randomly encounter this problem. The app worked fine 15 mins ago, I just reloaded it and it started crashing. Had no problems with amplify since I started the project 3 weeks ago.

<img width="782" alt="image" src="https://user-images.githubusercontent.com/102980855/217109479-99fc2eb5-5d92-4340-8e0c-7c30ec81ca02.png">

`W/AndroidKeysetManager(10498): 	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer$MainDataStream.finish(KeyStoreCryptoOperationChunkedStreamer.java:228)
W/AndroidKeysetManager(10498): 	at android.security.keystore2.KeyStoreCryptoOperationChunkedStreamer.doFinal(KeyStoreCryptoOperationChunkedStreamer.java:181)
W/AndroidKeysetManager(10498): 	at android.security.keystore2.AndroidKeyStoreAuthenticatedAESCipherSpi$BufferAllOutputUntilDoFinalStreamer.doFinal(AndroidKeyStoreAuthenticatedAESCipherSpi.java:396)
W/AndroidKeysetManager(10498): 	at android.security.keystore2.AndroidKeyStoreCipherSpiBase.engineDoFinal(AndroidKeyStoreCipherSpiBase.java:603)
W/AndroidKeysetManager(10498): 	... 24 more
I/flutter (10498): ERROR | CredentialStoreStateMachine | Emitted error: PlatformException(InvalidProtocolBufferException, com.google.crypto.tink.shaded.protobuf.InvalidProtocolBufferException: Protocol message contained an invalid tag (zero)., Cause: null, Stacktrace: com.google.crypto.tink.shaded.protobuf.InvalidProtocolBufferException: Protocol message contained an invalid tag (zero).
I/flutter (10498): 	at com.google.crypto.tink.shaded.protobuf.GeneratedMessageLite.parsePartialFrom(GeneratedMessageLite.java:1566)
I/flutter (10498): 	at com.google.crypto.tink.shaded.protobuf.GeneratedMessageLite.parseFrom(GeneratedMessageLite.java:1664)
I/flutter (10498): 	at com.google.crypto.tink.proto.Keyset.parseFrom(Keyset.java:957)
I/flutter (10498): 	at com.google.crypto.tink.integration.android.SharedPrefKeysetReader.read(SharedPrefKeysetReader.java:84)
I/flutter (10498): 	at com.google.crypto.tink.CleartextKeysetHandle.read(CleartextKeysetHandle.java:58)
I/flutter (10498): 	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.read(AndroidKeysetManager.java:328)
I/flutter (10498): 	at com.google.crypto.tink.integration.android.AndroidKeysetMa
I/flutter (10498): ERROR | HostedUiStateMachine | Emitted error: PlatformException(InvalidProtocolBufferException, com.google.crypto.tink.shaded.protobuf.InvalidProtocolBufferException: Protocol message contained an invalid tag (zero)., Cause: null, Stacktrace: com.google.crypto.tink.shaded.protobuf.InvalidProtocolBufferException: Protocol message contained an invalid tag (zero).
I/flutter (10498): 	at com.google.crypto.tink.shaded.protobuf.GeneratedMessageLite.parsePartialFrom(GeneratedMessageLite.java:1566)
I/flutter (10498): 	at com.google.crypto.tink.shaded.protobuf.GeneratedMessageLite.parseFrom(GeneratedMessageLite.java:1664)
I/flutter (10498): 	at com.google.crypto.tink.proto.Keyset.parseFrom(Keyset.java:957)
I/flutter (10498): 	at com.google.crypto.tink.integration.android.SharedPrefKeysetReader.read(SharedPrefKeysetReader.java:84)
I/flutter (10498): 	at com.google.crypto.tink.CleartextKeysetHandle.read(CleartextKeysetHandle.java:58)
I/flutter (10498): 	at com.google.crypto.tink.integration.android.AndroidKeysetManager$Builder.read(AndroidKeysetManager.java:328)
I/flutter (10498): 	at com.google.crypto.tink.integration.android.AndroidKeysetManager$B`

**Comment by Jordan-Nelson** [comment]
@alexandrudrosca - I'm working on a fix. Until then, I believe you can resolve this by opening App Info (long press on app icon) then Storage & Cache, and select clear storage. You may also have to force a backup of the device after that to clear the data from Androids cloud backup.

Uninstalling the app will not resolve this issue. The issue is caused by the data getting backed up to the cloud with Android backup. Uninstalling the app won't remove the cloud backup.

**Comment by alexandrudrosca** [comment]
Thank you for sharing the workaround. Luckily clearing the storage did the trick for now. Good luck with fixing it and BIG UP for contributing!

**Comment by asaarnak** [comment]
@Jordan-Nelson
I managed to get this error every time now on my phone, and can't login, even if i clear cache/data/uninstall app and restart phone i still get this error.

**Comment by Jordan-Nelson** [comment]
> even if i clear cache/data/uninstall app and restart phone i still get this error

What order are you performing those steps in?

If you uninstall and re-install the app, that will cause google to restore the backup version, which is what causes the issue. After you reinstall the app, you will have to clear the app data each time. 

If you do it in the reverse order (clear app data, then reinstall) I would expect you to see the error.

**Comment by landyrev** [comment]
Hey, @Jordan-Nelson! thanks for the PR and fix 🙌 

Do you have any plans on making a new release with this fix included?

**Comment by Jordan-Nelson** [comment]
Hey folks - I will see if we can get a release out in the next few days. I'll update this issue when the release is out.

**Comment by Jordan-Nelson** [comment]
The fix for this was included in v1.0.0-next.5. Please let us know if you have any issues after upgrading.

**Comment by Jordan-Nelson** [comment]
Closing this out since the fix has been released. Please open a new issue if you are facing any other troubles.
